lisa/sut_orchestrator/hyperv/get_assignable_devices.py

```diff
@@ -0,0 +1,302 @@
+    PROP_DEVICE_TYPE_PCI_EXPRESS_ENDPOINT = "2"
+    PROP_DEVICE_TYPE_PCI_EXPRESS_LEGACY_ENDPOINT = "3"
+    PROP_DEVICE_TYPE_PCI_EXPRESS_ROOT_COMPLEX_INTEGRATED_ENDPOINT = "4"
+    PROP_DEVICE_TYPE_PCI_EXPRESS_TREATED_AS_PCI = "5"
+    PROP_ACS_COMPATIBLE_UP_HIERARCHY_NOT_SUPPORTED = "0"
+    PROP_BASE_CLASS_DISPLAY_CTRL = "3"
+
+    def __init__(self, host_node: Node, log: Logger):
+        self.host_node = host_node
+        self.log = log
+        self.pwsh = self.host_node.tools[PowerShell]
+        self.pnp_allocated_resources: List[Dict[str, str]] = (
+            self.__load_pnp_allocated_resources()
+        )
+
+    def __get_devices_by_vendor_device_id(
+        self,
+        vendor_id: str,
+        device_id: str,
+    ) -> List[Dict[str, str]]:
+        """
+        Get the device ID list for given vendor/device ID combination
+        """
+        devices: List[Dict[str, str]] = []
+        device_regex = re.compile(
+            r"Description\s+:\s*(?P<desc>.+)\n.*DeviceID\s+:\s*(?P<device_id>.+)"
+        )
+
+        cmd = (
+            "Get-WmiObject Win32_PnPEntity -Filter "
+            f"\"DeviceID LIKE 'PCI\\\\VEN_{vendor_id}&DEV_{device_id}%'\""
+        )
+        stdout = self.pwsh.run_cmdlet(
+            cmdlet=cmd,
+            force_run=True,
+            sudo=True,
+        )
+
+        devices_str = stdout.strip().split("\r\n\r\n")
+        filtered_devices = [i.strip() for i in devices_str if i.strip() != ""]
+        for device_properties in filtered_devices:
+            res = find_group_in_lines(
+                lines=device_properties,
+                pattern=device_regex,
+                single_line=False,
+            )
+            if not res:
+                raise LisaException("Can not extract DeviceId/Description")
+
+            devices.append({
+                "device_id": res["device_id"].strip(),
+                "friendly_name": res["desc"].strip(),
+            })
+        return devices
+
+    def __get_pnp_device_property(self, device_id: str, property_name: str) -> str:
+        """
+        Retrieve a PnP device property by instance ID and property key.
+        """
+        cmd = (
+            "(Get-PnpDeviceProperty -InstanceId "
+            f"'{device_id}' '{property_name}').Data"
+        )
+
+        output = self.pwsh.run_cmdlet(
+            cmdlet=cmd,
+            sudo=True,
+            force_run=True,
+        )
+        return output.strip()
+
+    def __load_pnp_allocated_resources(self) -> List[Dict[str, str]]:
+        stdout = self.pwsh.run_cmdlet(
+            cmdlet="gwmi -query 'select * from Win32_PnPAllocatedResource'",
+            sudo=True,
+            force_run=True,
+        )
+        pnp_allocated_resources = stdout.strip().split("\r\n\r\n")
+        result: List[Dict[str, str]] = []
+        for rec in pnp_allocated_resources:
+            res: Dict[str, str] = {}
+            old_key: str = ""
+            for i in rec.strip().splitlines():
```

Comment:
Give some examples for output, so it's easier to compare if the output is expected. Use regexp is clearer to operate strings.

Reply:
@squirrelsc Done. Used named regex now. Added stdout for 2 records.